util/internal/test/common/src/test/java/org/hibernate/search/util/impl/test/reflect/ExperimentsTest.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.util.impl.test.reflect;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.hibernate.search.util.impl.test.extension.MultiRunExtension;
+
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class ExperimentsTest {
+
+	public static List<? extends Arguments> params() {
+		return Arrays.asList(
+				Arguments.of( "string1", 1, true ),
+				Arguments.of( "string2", 2, true ),
+				Arguments.of( "string3", 3, true )
+		);
+	}
+
+	@MultiRunExtension.EnvironmentTest(init = "init")
+	@MethodSource("params")
+	void env(String string, int number, boolean bool) {
```

Comment:
hmm, this `TestTemplateInvocationContextProvider` that I've tried to used is working with the test method, in this particular case it's any method annotated with `@EnvironmentTest`, and all it does is basically "providing a different context to this method to execute"... We can use it as a `@BeforeEach` . The execution of this `@MultiRunExtension.EnvironmentTest(init = "init") void env(String string, int number, boolean bool)` is performed by JUnit, and I just append a call to either init or to `@MultiRunExtension.TestToExecute` actual test method in the after-callback...

Reply:
And ideally we'd use `@Test` instead of `@MultiRunExtension.TestToExecute`, but I'm not sure how that could work exactly. Maybe the extension could prevent JUnit's default execution of `@Test` methods somehow? Not sure side-effects from one method to another are allowed, but it's worth having a look, at least.

Reply:
> I see you can add whatever extension you want in `TestTemplateInvocationContext#getAdditionalExtensions`, so maybe you could use `org.junit.jupiter.api.extension.ExecutionCondition` or `org.junit.jupiter.api.extension.InvocationInterceptor` to:
> 
>     1. Skip some executions of the `setup` method.
> 
>     2. Replace some executions of the `setup` method with the actual tests.

From your other experiment, I take it this didn't work out?

Reply:
😔 nope... it didn't 🥲

Reply:
The additional extensions didn't work, or... ?

Reply:
See the new variant 😄 after our discussion 😉

Reply:
Nice improvement, thanks!

I didn't look at the code yet, but I'll make a few comments on the API.

About the cleanup, I'd suggest in particular:

* Renaming `@MultiRunExtension.EnvironmentTest` to `@ParameterizedSetup`
* Using `@Test` instead of `@MultiRunExtension.TestToExecute` if possible (see below), or failing that renaming it to`@ParameterizedSetup.Test`.

About the behavior, two questions:

1. The "env" (setup) method will likely need to set fields that will have to be read by test methods, which I expect will require using `@TestInstance(Lifecycle.PER_CLASS)`. Could you check this, and if it is indeed required, could you either:
   a. Find a way to force the PER_CLASS lifecycle as soon as `MultiRunExtension` is used
   b. OR (if the above is not possible) check the test lifecycle in `MultiRunExtension` and throw an exception if it's `PER_METHOD` (e.g. "this extension requires test instances to be shared across test methods; apply `@TEstInstance(Lifecycle.PER_CLASS)` to your test").
2. Did you look into using simply `@Test`/`@ParameterizedTest` for the test methods? If that's not possible from the extension on the `env` method, you could try moving the extension annotation to the class level, so that it applies extensions to all test methods, effectively removing all test methods from the testing plan, and restoring them as part of the setup?

Reply:
ooook, I think you can now look at the `ParameterizedClassExtensionTest` to see how it can look like, and even work 😆.
There is a marker annotation `@ParameterizedClass` beside the fact that it has a `@TestInstance(Lifecycle.PER_CLASS)` to be able to correctly access context and prevent some weird behavior it also serves as a marker for a filter `ParameterizedClassPostDiscoveryFilter`. This filter is there to tell JUnit to ignore `@Test`/`@ParameterizedTest` if we are working with a parameterized class.
Then there is a `@ParameterizedSetup` that we put on the setup method along side some of the argument source annotations.
Then you can have `@Test`/`@ParameterizedTest`  in the class and the extension will discover and execute them ...

Reply:
Well API-wise, this looks perfect. I'll add comments about various additional checks that could make sense and will try to finish my review.
Thanks a lot!